### PR TITLE
JSONSerialization for Info structures

### PR DESCRIPTION
**Pull request by bucketzxm**

### CODE REVIEW

File: include/openrave/json.h [hunk 1]
```diff
@@ -347,6 +363,20 @@ inline void RaveDeserializeJSON(const rapidjson::Value &value, std::map<K, V>& m
...
 }
 
+template <typename T>
+inline void RaveDeserializeJSON(const rapidjson::Value& value, std::set<T>& s)
+{
+    if (!value.IsArray()) {
+        throw OPENRAVE_EXCEPTION_FORMAT0("failed to deserialize json, value cannot be decoded as a set", ORE_InvalidArguments);
+    }
+    s.clear();
+    for(rapidjson::SizeType i=0; i<value.Size(); ++i) {
```
Comment (ziyan) [inline include/openrave/json.h]: should iterate with value.Begin() and value.End()

File: test/test_json.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+class TestJSONSerialization(EnvironmentSetup):
+    _name_ = 'TestJSONSerializeJSON'
+    _daefiles = None
+
+    def __init__(self, *args, **kwargs):
+        super(TestJSONSerialization, self).__init__(*args, **kwargs)
+
+    def setup(self):
+        super(TestJSONSerialization, self).setup()
+        self._daefiles = []
```
Comment (bucketzxm) [inline test/test_json.py]: use set

File: include/openrave/kinbody.h [hunk 3]
```diff
@@ -249,6 +252,9 @@ class OPENRAVE_API KinBody : public InterfaceBase
         LinkInfo(const LinkInfo& other);
         LinkInfo& operator=(const LinkInfo& other);
 
+        virtual void SerializeJSON(rapidjson::Value &value, rapidjson::Document::AllocatorType& allocator, int options=0);
```
Comment (bucketzxm) [inline include/openrave/kinbody.h]: add const

**Comment by rdiankov** [comment]
@bucketzxm There were several problems:

1. Please use `Reserve` for all rapidjson Array usages. 
2. Missed calling `geometriesValue.SetArray();` in `LinkInfo::SerializeJSON`, so geometries were not being serialized.

**Pull request merged**